File: src/streaming_preview.cpp
#include "streaming_preview.h"
#include "loguru.hpp"
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <cstdlib>
#include <cstring>

// ============================================================================
// GzipTransform implementation
// ============================================================================

GzipTransform::GzipTransform() {
    memset(&m_zstream, 0, sizeof(m_zstream));

    // 16 + MAX_WBITS tells zlib to detect gzip or zlib format automatically
    int ret = inflateInit2(&m_zstream, 16 + MAX_WBITS);
    if (ret != Z_OK) {
        LOG_F(ERROR, "GzipTransform: inflateInit2 failed with code %d", ret);
        m_error = true;
        return;
    }

    m_initialized = true;
    LOG_F(INFO, "GzipTransform: initialized successfully");
}

GzipTransform::~GzipTransform() {
    if (m_initialized) {
        inflateEnd(&m_zstream);
        m_initialized = false;
    }
}

std::string GzipTransform::transform(const char* data, size_t len) {
    if (!m_initialized || m_error || len == 0) {
        return "";
    }

    std::string output;
    output.reserve(len * 2);  // Guess: decompressed is ~2x compressed

    // Set up input
    m_zstream.next_in = reinterpret_cast<Bytef*>(const_cast<char*>(data));
    m_zstream.avail_in = static_cast<uInt>(len);

    // Decompress in chunks
    char outbuf[32768];  // 32KB output buffer

    do {
        m_zstream.next_out = reinterpret_cast<Bytef*>(outbuf);
        m_zstream.avail_out = sizeof(outbuf);

        int ret = inflate(&m_zstream, Z_NO_FLUSH);

        if (ret == Z_STREAM_ERROR) {
            LOG_F(ERROR, "GzipTransform: Z_STREAM_ERROR");
            m_error = true;
            return output;
        }

        switch (ret) {
            case Z_NEED_DICT:
            case Z_DATA_ERROR:
            case Z_MEM_ERROR:
                LOG_F(ERROR, "GzipTransform: inflate error %d: %s",
                      ret, m_zstream.msg ? m_zstream.msg : "unknown");
                m_error = true;
                return output;
        }

        size_t have = sizeof(outbuf) - m_zstream.avail_out;
        output.append(outbuf, have);

        // Z_STREAM_END means we've finished decompressing
        if (ret == Z_STREAM_END) {
            LOG_F(INFO, "GzipTransform: reached end of compressed stream");
            break;
        }

    } while (m_zstream.avail_out == 0);

    return output;
}

std::string GzipTransform::flush() {
    if (!m_initialized || m_error) {
        return "";
    }

    std::string output;
    char outbuf[32768];

    // Try to flush any remaining data
    m_zstream.next_in = nullptr;
    m_zstream.avail_in = 0;

    do {
        m_zstream.next_out = reinterpret_cast<Bytef*>(outbuf);
        m_zstream.avail_out = sizeof(outbuf);

        int ret = inflate(&m_zstream, Z_FINISH);

        if (ret == Z_STREAM_ERROR) {
            LOG_F(ERROR, "GzipTransform::flush: Z_STREAM_ERROR");
            m_error = true;
            break;
        }

        size_t have = sizeof(outbuf) - m_zstream.avail_out;
        if (have > 0) {
            output.append(outbuf, have);
        }

        if (ret == Z_STREAM_END) {
            break;
        }

        // Z_BUF_ERROR is OK here - just means no more output available
        if (ret == Z_BUF_ERROR) {
            break;
        }

    } while (m_zstream.avail_out == 0);

    LOG_F(INFO, "GzipTransform::flush: produced %zu bytes", output.size());
    return output;
}

// ============================================================================
// ZstdTransform implementation
// ============================================================================

ZstdTransform::ZstdTransform() {
    m_dstream = ZSTD_createDStream();
    if (!m_dstream) {
        LOG_F(ERROR, "ZstdTransform: ZSTD_createDStream failed");
        m_error = true;
        return;
    }

    size_t ret = ZSTD_initDStream(m_dstream);
    if (ZSTD_isError(ret)) {
        LOG_F(ERROR, "ZstdTransform: ZSTD_initDStream failed: %s", ZSTD_getErrorName(ret));
        m_error = true;
        return;
    }

    LOG_F(INFO, "ZstdTransform: initialized successfully");
}

ZstdTransform::~ZstdTransform() {
    if (m_dstream) {
        ZSTD_freeDStream(m_dstream);
        m_dstream = nullptr;
    }
}

std::string ZstdTransform::transform(const char* data, size_t len) {
    if (!m_dstream || m_error || len == 0) {
        return "";
    }

    std::string output;
    output.reserve(len * 4);  // Guess: decompressed is ~4x compressed

    ZSTD_inBuffer input = { data, len, 0 };

    // Decompress in chunks
    const size_t outBufSize = ZSTD_DStreamOutSize();
    std::vector<char> outbuf(outBufSize);

    while (input.pos < input.size) {
        ZSTD_outBuffer outBuffer = { outbuf.data(), outbuf.size(), 0 };

        size_t ret = ZSTD_decompressStream(m_dstream, &outBuffer, &input);

        if (ZSTD_isError(ret)) {
            LOG_F(ERROR, "ZstdTransform: ZSTD_decompressStream error: %s", ZSTD_getErrorName(ret));
            m_error = true;
            return output;
        }

        if (outBuffer.pos > 0) {
            output.append(outbuf.data(), outBuffer.pos);
        }

        // ret == 0 means end of frame, but there might be more frames
        // Continue processing input
    }

    return output;
}

std::string ZstdTransform::flush() {
    // For zstd streaming, all data should be flushed during transform()
    // This is called when stream completes, nothing special needed
    if (!m_dstream || m_error) {
        return "";
    }

    LOG_F(INFO, "ZstdTransform::flush: stream complete");
    return "";
}

// ============================================================================
// StreamingFilePreview implementation
// ============================================================================

StreamingFilePreview::StreamingFilePreview(
    const std::string& bucket,
    const std::string& key,
    const std::string& initialData,
    size_t totalFileSize,
    std::unique_ptr<IStreamTransform> transform)
    : m_bucket(bucket)
    , m_key(key)
    , m_totalSourceSize(totalFileSize)
    , m_transform(transform ? std::move(transform) : std::make_unique<PassThroughTransform>())
{
    // Create temp file
    const char* tmpdir = std::getenv("TMPDIR");
    if (!tmpdir) tmpdir = "/tmp";

    m_tempFilePath = std::string(tmpdir) + "/s6ui_preview_XXXXXX";

    // mkstemp modifies the string in place
    std::vector<char> pathBuf(m_tempFilePath.begin(), m_tempFilePath.end());
    pathBuf.push_back('\0');

    m_fd = mkstemp(pathBuf.data());
    if (m_fd < 0) {
        LOG_F(ERROR, "StreamingFilePreview: failed to create temp file: %s", strerror(errno));
        return;
    }

    m_tempFilePath = pathBuf.data();
    LOG_F(INFO, "StreamingFilePreview: created temp file %s for %s/%s (total=%zu bytes)",
          m_tempFilePath.c_str(), bucket.c_str(), key.c_str(), totalFileSize);

    // Initialize with first line offset
    m_lineOffsets.push_back(0);

    // Write initial data
    if (!initialData.empty()) {
        std::string transformed = m_transform->transform(initialData.data(), initialData.size());
        writeToTempFile(transformed.data(), transformed.size());
        m_bytesDownloaded = initialData.size();

        // Check if we got the whole file
        if (m_bytesDownloaded >= m_totalSourceSize) {
            finishStream();
        }
    }
}

StreamingFilePreview::~StreamingFilePreview() {
    if (m_fd >= 0) {
        close(m_fd);
        m_fd = -1;
    }

    // Delete temp file
    if (!m_tempFilePath.empty()) {
        LOG_F(INFO, "StreamingFilePreview: deleting temp file %s", m_tempFilePath.c_str());
        unlink(m_tempFilePath.c_str());
    }
}

void StreamingFilePreview::setTransform(std::unique_ptr<IStreamTransform> transform) {
    std::lock_guard<std::mutex> lock(m_mutex);
    if (m_bytesDownloaded > 0) {
        LOG_F(WARNING, "StreamingFilePreview: setTransform called after data received, ignoring");
        return;
    }
    m_transform = std::move(transform);
}

void StreamingFilePreview::appendChunk(const std::string& data, size_t offset) {
    std::lock_guard<std::mutex> lock(m_mutex);

    if (m_fd < 0) {
        LOG_F(WARNING, "StreamingFilePreview: appendChunk called but no temp file");
        return;
    }

    if (offset != m_bytesDownloaded) {
        LOG_F(WARNING, "StreamingFilePreview: chunk offset mismatch, expected %zu got %zu",
              m_bytesDownloaded, offset);
        // We could handle out-of-order chunks here, but for now assume sequential
        return;
    }

    // Transform the data (e.g., decompress)
    std::string transformed = m_transform->transform(data.data(), data.size());

    // Write to temp file
    writeToTempFile(transformed.data(), transformed.size());
    m_bytesDownloaded += data.size();

    LOG_F(1, "StreamingFilePreview: appended %zu bytes at offset %zu, total downloaded=%zu/%zu, lines=%zu",
          data.size(), offset, m_bytesDownloaded, m_totalSourceSize, m_lineOffsets.size());

    // Check if we're done
    if (m_bytesDownloaded >= m_totalSourceSize) {
        finishStream();
    }
}

void StreamingFilePreview::finishStream() {
    // Already holding lock from caller or constructor

    if (m_complete) return;

    // Flush any remaining transform data
    std::string remaining = m_transform->flush();
    if (!remaining.empty()) {
        writeToTempFile(remaining.data(), remaining.size());
    }

    m_complete = true;
    LOG_F(INFO, "StreamingFilePreview: stream complete, %zu bytes downloaded, %zu bytes written, %zu lines",
          m_bytesDownloaded, m_bytesWritten, m_lineOffsets.size());
}

void StreamingFilePreview::writeToTempFile(const char* data, size_t len) {
    // Caller must hold lock

    if (m_fd < 0 || len == 0) return;

    // Write data
    ssize_t written = write(m_fd, data, len);
    if (written < 0) {
        LOG_F(ERROR, "StreamingFilePreview: write failed: %s", strerror(errno));
        return;
    }

    // Index newlines in the data we just wrote
    size_t baseOffset = m_bytesWritten;
    indexNewlines(data, len, baseOffset);

    m_bytesWritten += static_cast<size_t>(written);
}

void StreamingFilePreview::indexNewlines(const char* data, size_t len, size_t baseOffset) {
    // Caller must hold lock

    for (size_t i = 0; i < len; ++i) {
        if (data[i] == '\n') {
            // The next line starts at the byte after this newline
            size_t nextLineOffset = baseOffset + i + 1;
            // Only add if there's more data (or will be more data)
            if (nextLineOffset < m_bytesWritten + len || !m_complete) {
                m_lineOffsets.push_back(nextLineOffset);
            }
        }
    }
}

size_t StreamingFilePreview::lineCount() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_lineOffsets.size();
}

size_t StreamingFilePreview::bytesDownloaded() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_bytesDownloaded;
}

size_t StreamingFilePreview::bytesWritten() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_bytesWritten;
}

size_t StreamingFilePreview::totalSourceBytes() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_totalSourceSize;
}

bool StreamingFilePreview::isComplete() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_complete;
}

size_t StreamingFilePreview::nextByteNeeded() const {
    std::lock_guard<std::mutex> lock(m_mutex);
    return m_bytesDownloaded;
}

std::string StreamingFilePreview::getLine(size_t lineIndex) const {
    std::lock_guard<std::mutex> lock(m_mutex);

    if (m_fd < 0) return "";
    if (lineIndex >= m_lineOffsets.size()) return "";

    size_t startOffset = m_lineOffsets[lineIndex];

    // Determine end offset
    size_t endOffset;
    if (lineIndex + 1 < m_lineOffsets.size()) {
        // End at start of next line (minus the newline character)
        endOffset = m_lineOffsets[lineIndex + 1] - 1;
    } else {
        // Last line - end at current write position
        endOffset = m_bytesWritten;
    }

    if (startOffset >= endOffset) return "";

    size_t lineLen = endOffset - startOffset;

    // Cap line length to prevent huge allocations
    constexpr size_t MAX_LINE_LEN = 10 * 1024 * 1024;  // 10MB max line
    if (lineLen > MAX_LINE_LEN) {
        lineLen = MAX_LINE_LEN;
    }

    // Read the line from temp file
    std::string line(lineLen, '\0');

    ssize_t bytesRead = pread(m_fd, line.data(), lineLen, static_cast<off_t>(startOffset));
    if (bytesRead < 0) {
        LOG_F(ERROR, "StreamingFilePreview: pread failed: %s", strerror(errno));
        return "";
    }

    line.resize(static_cast<size_t>(bytesRead));

    // Trim trailing newline/carriage return if present
    while (!line.empty() && (line.back() == '\n' || line.back() == '\r')) {
        line.pop_back();
    }

    return line;
}

bool StreamingFilePreview::isLineComplete(size_t lineIndex) const {
    std::lock_guard<std::mutex> lock(m_mutex);

    if (lineIndex >= m_lineOffsets.size()) return false;

    // Line is complete if:
    // 1. There's a next line (meaning we found a newline after this line)
    // 2. OR the file download is complete
    if (lineIndex + 1 < m_lineOffsets.size()) {
        return true;  // There's a next line, so this one is terminated
    }

    // This is the last known line - it's only complete if file is fully downloaded
    return m_complete;
}

std::string StreamingFilePreview::getAllContent() const {
    std::lock_guard<std::mutex> lock(m_mutex);

    if (m_fd < 0 || m_bytesWritten == 0) {
        return "";
    }

    std::string content(m_bytesWritten, '\0');

    ssize_t bytesRead = pread(m_fd, content.data(), m_bytesWritten, 0);
    if (bytesRead < 0) {
        LOG_F(ERROR, "StreamingFilePreview::getAllContent: pread failed: %s", strerror(errno));
        return "";
    }

    content.resize(static_cast<size_t>(bytesRead));
    return content;
}

File: rust/src/streaming_preview.rs
use std::fs;
use std::io::{Read, Write};
use std::sync::Mutex;

#[cfg(unix)]
use std::os::unix::fs::FileExt;

/// Abstract interface for data transformation (decompression, etc.)
pub trait StreamTransform: Send {
    fn transform(&mut self, data: &[u8]) -> Vec<u8>;
    fn flush(&mut self) -> Vec<u8>;
}

/// Shared transform/flush logic for buffered decompression transforms.
/// Buffers all compressed data and re-decompresses from scratch each time,
/// emitting only the newly decompressed bytes.
macro_rules! impl_buffered_transform {
    ($ty:ty) => {
        impl StreamTransform for $ty {
            fn transform(&mut self, data: &[u8]) -> Vec<u8> {
                if self.error || data.is_empty() {
                    return vec![];
                }
                self.buffer.extend_from_slice(data);
                let all = self.decompress_all();
                if all.len() > self.emitted {
                    let new_data = all[self.emitted..].to_vec();
                    self.emitted = all.len();
                    new_data
                } else {
                    vec![]
                }
            }

            fn flush(&mut self) -> Vec<u8> {
                if self.error {
                    return vec![];
                }
                let all = self.decompress_all();
                if all.len() > self.emitted {
                    let new_data = all[self.emitted..].to_vec();
                    self.emitted = all.len();
                    new_data
                } else {
                    vec![]
                }
            }
        }
    };
}

/// Pass-through transform (no transformation)
pub struct PassThroughTransform;

impl StreamTransform for PassThroughTransform {
    fn transform(&mut self, data: &[u8]) -> Vec<u8> {
        data.to_vec()
    }
    fn flush(&mut self) -> Vec<u8> {
        vec![]
    }
}

/// Gzip decompression transform.
/// Buffers all compressed data and re-decompresses from scratch each time,
/// tracking how many decompressed bytes were already emitted.
pub struct GzipTransform {
    buffer: Vec<u8>,
    emitted: usize,
    error: bool,
}

impl GzipTransform {
    pub fn new() -> Self {
        Self {
            buffer: Vec::new(),
            emitted: 0,
            error: false,
        }
    }

    pub fn has_error(&self) -> bool {
        self.error
    }

    fn decompress_all(&self) -> Vec<u8> {
        use flate2::read::GzDecoder;
        let cursor = std::io::Cursor::new(&self.buffer);
        let mut decoder = GzDecoder::new(cursor);
        let mut output = Vec::new();
        let _ = decoder.read_to_end(&mut output);
        output
    }
}

impl_buffered_transform!(GzipTransform);

/// Zstd decompression transform.
/// Same approach: buffer all compressed data, decompress from scratch.
pub struct ZstdTransform {
    buffer: Vec<u8>,
    emitted: usize,
    error: bool,
}

impl ZstdTransform {
    pub fn new() -> Self {
        Self {
            buffer: Vec::new(),
            emitted: 0,
            error: false,
        }
    }

    pub fn has_error(&self) -> bool {
        self.error
    }

    fn decompress_all(&self) -> Vec<u8> {
        let cursor = std::io::Cursor::new(&self.buffer);
        let mut decoder = match zstd::stream::Decoder::new(cursor) {
            Ok(d) => d,
            Err(_) => return vec![],
        };
        let mut output = Vec::new();
        let _ = decoder.read_to_end(&mut output);
        output
    }
}

impl_buffered_transform!(ZstdTransform);

/// Manages streaming download of a file to a temp file with newline indexing.
pub struct StreamingFilePreview {
    bucket: String,
    key: String,
    temp_file_path: String,
    file: Option<fs::File>,
    total_source_size: usize,
    inner: Mutex<StreamingInner>,
}

struct StreamingInner {
    bytes_downloaded: usize,
    bytes_written: usize,
    complete: bool,
    line_offsets: Vec<usize>,
    transform: Box<dyn StreamTransform>,
}

impl StreamingFilePreview {
    pub fn new(
        bucket: String,
        key: String,
        initial_data: &[u8],
        total_file_size: usize,
        transform: Option<Box<dyn StreamTransform>>,
    ) -> Self {
        let tmpdir = std::env::var("TMPDIR").unwrap_or_else(|_| "/tmp".to_string());
        use std::sync::atomic::{AtomicU64, Ordering as AtomicOrdering};
        static COUNTER: AtomicU64 = AtomicU64::new(0);
        let temp_file_path = format!(
            "{}/s6ui_preview_{}_{}",
            tmpdir,
            std::process::id(),
            COUNTER.fetch_add(1, AtomicOrdering::Relaxed)
        );

        let file = fs::OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .truncate(true)
            .open(&temp_file_path)
            .ok();

        let transform = transform.unwrap_or_else(|| Box::new(PassThroughTransform));

        let mut preview = StreamingFilePreview {
            bucket,
            key,
            temp_file_path,
            file,
            total_source_size: total_file_size,
            inner: Mutex::new(StreamingInner {
                bytes_downloaded: 0,
                bytes_written: 0,
                complete: false,
                line_offsets: vec![0],
                transform,
            }),
        };

        // Write initial data
        if !initial_data.is_empty() {
            let mut inner = preview.inner.lock().unwrap();
            let transformed = inner.transform.transform(initial_data);
            if !transformed.is_empty() {
                if let Some(ref mut f) = preview.file {
                    let base_offset = inner.bytes_written;
                    let _ = f.write_all(&transformed);
                    Self::index_newlines(&transformed, base_offset, &mut inner.line_offsets);
                    inner.bytes_written += transformed.len();
                }
            }
            inner.bytes_downloaded = initial_data.len();

            if inner.bytes_downloaded >= total_file_size {
                Self::finish_stream_inner(&mut inner, &mut preview.file);
            }
        }

        preview
    }

    fn finish_stream_inner(inner: &mut StreamingInner, file: &mut Option<fs::File>) {
        if inner.complete {
            return;
        }

        let remaining = inner.transform.flush();
        if !remaining.is_empty() {
            if let Some(f) = file {
                let base_offset = inner.bytes_written;
                let _ = f.write_all(&remaining);
                Self::index_newlines(&remaining, base_offset, &mut inner.line_offsets);
                inner.bytes_written += remaining.len();
            }
        }

        inner.complete = true;
    }

    fn index_newlines(data: &[u8], base_offset: usize, line_offsets: &mut Vec<usize>) {
        for (i, &b) in data.iter().enumerate() {
            if b == b'\n' {
                line_offsets.push(base_offset + i + 1);
            }
        }
    }

    /// Append a new chunk from streaming download.
    pub fn append_chunk(&self, data: &[u8], offset: usize) {
        let mut inner = self.inner.lock().unwrap();

        if self.file.is_none() {
            return;
        }

        if offset != inner.bytes_downloaded {
            eprintln!(
                "StreamingFilePreview: chunk offset mismatch, expected {} got {}",
                inner.bytes_downloaded, offset
            );
            return;
        }

        let transformed = inner.transform.transform(data);
        if !transformed.is_empty() {
            // We need to write to the file but self.file is behind &self.
            // Use the file handle via unsafe raw fd for pwrite, or just accept the limitation.
            // Actually, File::write_all needs &mut File, but we can use write on &File on unix.

            if let Some(ref f) = self.file {
                let base_offset = inner.bytes_written;
                let _ = f.write_all_at(&transformed, base_offset as u64);
                Self::index_newlines(&transformed, base_offset, &mut inner.line_offsets);
                inner.bytes_written += transformed.len();
            }
        }
        inner.bytes_downloaded += data.len();

        if inner.bytes_downloaded >= self.total_source_size {
            // For finish_stream, we can't pass &mut self.file since self is &self.
            // Instead, do an inline finish.
            if !inner.complete {
                let remaining = inner.transform.flush();
                if !remaining.is_empty() {
        
                    if let Some(ref f) = self.file {
                        let base_offset = inner.bytes_written;
                        let _ = f.write_all_at(&remaining, base_offset as u64);
                        Self::index_newlines(&remaining, base_offset, &mut inner.line_offsets);
                        inner.bytes_written += remaining.len();
                    }
                }
                inner.complete = true;
            }
        }
    }

    pub fn line_count(&self) -> usize {
        self.inner.lock().unwrap().line_offsets.len()
    }

    pub fn bytes_downloaded(&self) -> usize {
        self.inner.lock().unwrap().bytes_downloaded
    }

    pub fn bytes_written(&self) -> usize {
        self.inner.lock().unwrap().bytes_written
    }

    pub fn total_source_bytes(&self) -> usize {
        self.total_source_size
    }

    pub fn is_complete(&self) -> bool {
        self.inner.lock().unwrap().complete
    }

    pub fn next_byte_needed(&self) -> usize {
        self.inner.lock().unwrap().bytes_downloaded
    }

    /// Get a specific line (0-indexed).
    pub fn get_line(&self, line_index: usize) -> String {
        let inner = self.inner.lock().unwrap();

        if line_index >= inner.line_offsets.len() {
            return String::new();
        }

        let start_offset = inner.line_offsets[line_index];
        let end_offset = if line_index + 1 < inner.line_offsets.len() {
            inner.line_offsets[line_index + 1].saturating_sub(1)
        } else {
            inner.bytes_written
        };

        let end_offset = end_offset.min(inner.bytes_written);
        if start_offset >= end_offset {
            return String::new();
        }

        let line_len = (end_offset - start_offset).min(10 * 1024 * 1024);
        drop(inner);


        if let Some(ref f) = self.file {
            let mut buf = vec![0u8; line_len];
            match f.read_at(&mut buf, start_offset as u64) {
                Ok(n) => {
                    buf.truncate(n);
                    while buf.last() == Some(&b'\n') || buf.last() == Some(&b'\r') {
                        buf.pop();
                    }
                    String::from_utf8_lossy(&buf).into_owned()
                }
                Err(_) => String::new(),
            }
        } else {
            String::new()
        }
    }

    /// Get all content written so far.
    pub fn get_all_content(&self) -> Vec<u8> {
        let inner = self.inner.lock().unwrap();
        let bytes_written = inner.bytes_written;
        drop(inner);

        if bytes_written == 0 {
            return vec![];
        }


        if let Some(ref f) = self.file {
            let mut buf = vec![0u8; bytes_written];
            match f.read_at(&mut buf, 0) {
                Ok(n) => {
                    buf.truncate(n);
                    buf
                }
                Err(_) => vec![],
            }
        } else {
            vec![]
        }
    }

    pub fn is_line_complete(&self, line_index: usize) -> bool {
        let inner = self.inner.lock().unwrap();
        if line_index >= inner.line_offsets.len() {
            return false;
        }
        if line_index + 1 < inner.line_offsets.len() {
            return true;
        }
        inner.complete
    }

    pub fn bucket(&self) -> &str {
        &self.bucket
    }

    pub fn key(&self) -> &str {
        &self.key
    }

    pub fn temp_file_path(&self) -> &str {
        &self.temp_file_path
    }
}

impl Drop for StreamingFilePreview {
    fn drop(&mut self) {
        // Drop the file handle before removing
        drop(self.file.take());
        if !self.temp_file_path.is_empty() {
            let _ = fs::remove_file(&self.temp_file_path);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_streaming_preview_basic() {
        let data = b"line1\nline2\nline3\n";
        let preview = StreamingFilePreview::new(
            "bucket".to_string(),
            "key".to_string(),
            data,
            data.len(),
            None,
        );

        assert_eq!(preview.line_count(), 4); // 3 lines + empty after last \n
        assert_eq!(preview.get_line(0), "line1");
        assert_eq!(preview.get_line(1), "line2");
        assert_eq!(preview.get_line(2), "line3");
        assert!(preview.is_complete());
    }

    #[test]
    fn test_streaming_preview_chunked() {
        let total = b"hello\nworld\n";
        let preview = StreamingFilePreview::new(
            "bucket".to_string(),
            "key".to_string(),
            &total[..6], // "hello\n"
            total.len(),
            None,
        );

        assert_eq!(preview.line_count(), 2);
        assert_eq!(preview.get_line(0), "hello");
        assert!(!preview.is_complete());

        preview.append_chunk(&total[6..], 6);
        assert_eq!(preview.line_count(), 3);
        assert_eq!(preview.get_line(1), "world");
        assert!(preview.is_complete());
    }
}
